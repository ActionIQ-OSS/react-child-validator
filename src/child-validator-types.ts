import * as React from "react";

export type FunctionType = { (el: ReactElement): boolean, name: string };
export type ReactChild = React.ReactChild;
export type ReactClass = React.ComponentClass;
export type ReactElement = React.ReactElement<any>;
export type ChildGroupMatchReturn = ReactElement | ReactElement[];

export type Mapper = (_: ReactElement, _2: any) => ReactElement;
export type SingleChildType = ReactElement | ReactClass | FunctionType | string;
export type HasDisplayName = { displayName: string }

export interface IChildMatch {
  types: SingleChildType[];
  min: number;
  max: number | undefined;
}

export interface IChildGroup {
  match: IChildMatch;
  mapper?: Mapper;
}
