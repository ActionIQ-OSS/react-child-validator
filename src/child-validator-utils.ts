import {
  IChildGroup,
  IChildMatch,
  FunctionType,
  Mapper,
  SingleChildType,
  ReactElement,
} from "./child-validator-types";

/*
 * Check if the received child type matches one of the expected types
 */
export function validateChild(
  expectedTypes: SingleChildType[],
  receivedChild: ReactElement,
): boolean {
  const receivedType = receivedChild.type;
  const found = expectedTypes.find(t => {
    const isDomElement = typeof t === "string";
    const isComponent = !!((t as any)["displayName"]);
    const isFunction = typeof t === "function" && !(t as any)["displayName"];

    return (
      (isDomElement && t === receivedType)
      || (isFunction && (t as FunctionType)(receivedChild))
      || (isComponent && (t as any) === receivedType)
    );
  })
  return found !== undefined;
}

export const zeroOrOneOf = (...args: SingleChildType[]): IChildGroup => ({
  match: {
    types: args,
    min: 0,
    max: 1,
  },
});

export const zeroOrMoreOf = (...args: SingleChildType[]): IChildGroup => ({
  match: {
    types: args,
    min: 0,
    max: undefined,
  },
});

export const oneOf = (...args: SingleChildType[]): IChildGroup => ({
  match: {
    types: args,
    min: 1,
    max: 1,
  },
});

export const oneOrMoreOf = (...args: SingleChildType[]): IChildGroup => ({
  match: {
    types: args,
    min: 1,
    max: undefined,
  },
});

export const countOf = (count: number, ...args: SingleChildType[]): IChildGroup => ({
  match: {
    types: args,
    min: count,
    max: count,
  },
});

export const countOrMoreOf = (count: number, ...args: SingleChildType[]): IChildGroup => ({
  match: {
    types: args,
    min: count,
    max: undefined,
  },
});

export const countOrLessOf = (count: number, ...args: SingleChildType[]): IChildGroup => ({
  match: {
    types: args,
    min: 0,
    max: count,
  },
});

export const countBetweenOf = (min: number, max: number, ...args: SingleChildType[]): IChildGroup => ({
  match: {
    types: args,
    min,
    max,
  },
});

export const withMapper = (childGroup: IChildGroup, mapper: Mapper): IChildGroup => ({
  ...childGroup,
  mapper,
})