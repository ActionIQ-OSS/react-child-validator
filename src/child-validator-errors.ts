import { IChildGroup, FunctionType, HasDisplayName, SingleChildType, ReactClass, ReactElement } from "./child-validator-types";

/*
 * Given a SingleChildType[], build a human readable error message. If a component
 * print the displayName, if a function print the name, if fn doesn't have a name
 * toString it.
 */
const errorStringTypes = (types: SingleChildType[]) => {
  const generateString = (t: SingleChildType) => {
    return (t as ReactClass).displayName || (t as FunctionType).name || t.toString()
  }

  if (types.length === 1) {
    return generateString(types[0]);
  }
  return "[" + types.map(t => generateString(t)).join(", ") + "]";
}

/*
 * Given a ReactElement, build a (mostly) human readable version of it
 */
const errorStringChild = (child: ReactElement) => {
  const props = child.props;
  const keys  = Object.keys(props);
  const propString = keys.map(k => `${k}={${props[k].toString()}}`).join(" ")
 
  return `<${(child.type as ReactClass).displayName}${propString ? " " : ""}${propString} />`;
}

/*
 * Checks if there are extra children that weren't matched, throws an error if true
 */
export function checkUnmatchedChildren(
  componentClass: HasDisplayName,
  nextChildIndex: number,
  numChildren: number,
  throwError: boolean,
) {
  if (nextChildIndex >= numChildren || !throwError) {
    return;
  }

  throw new Error(
    `Too many children passed to <${componentClass.displayName} />. Matched ${nextChildIndex} of ${numChildren}.`,
  );
}

/*
 * Checks if too few children or two many children were given for a specific ChildGroup.
 * Throws the approriate error with as much details as possible.
 */
export function checkIncorrectChildrenCountError(
  componentClass: HasDisplayName,
  childGroup: IChildGroup,
  numMatched: number,
  currentChild: ReactElement | undefined,
  throwError: boolean,
) {
  if (!throwError) {
    return;
  }

  const { min, max, types } = childGroup.match;
  if (numMatched >= min && (max === undefined || numMatched <= max)) {
    return;
  }

  const isTooFew = numMatched < min;
  const fewOrMany = isTooFew ? "Too few" : "Too many";

  let atBlank = "";
  let atCount = "";
  if (max !== undefined && max !== min) {
    atBlank = "between ";
    atCount = `[${min}, ${max}]`;
  } else {
    if (isTooFew) {
      atBlank = "at least ";
      atCount = min + "";
    } else {
      atBlank = "at most ";
      atCount = max + "";
    }
  }
  if (max === min) {
    atBlank = "";
  }

  throw new Error(
    `${fewOrMany} children of type ${errorStringTypes(types)} for <${componentClass.displayName} />.`
    + ` Expected ${atBlank}${atCount}, Received: ${numMatched}.`
    + (currentChild ? ` Failed validation on ${errorStringChild(currentChild)}.` : ""),
  );
}
