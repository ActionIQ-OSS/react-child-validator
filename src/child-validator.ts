import * as React from "react";

import { checkUnmatchedChildren, checkIncorrectChildrenCountError } from "./child-validator-errors";
import {
  ChildGroupMatchReturn,
  IChildGroup,
  HasDisplayName,
  Mapper,
  ReactChild,
  ReactElement,
} from "./child-validator-types";
import { validateChild } from "./child-validator-utils";

export class ReactChildValidator {
  /* children to validate */
  private children: ReactElement[];
  /* groups to validate against */
  private groups: IChildGroup[];
  /* parent component's class */
  private componentClass: HasDisplayName;
  /* props of the parent component */
  private props: Object;
  /* if should throw errors if they occur */
  private throwError: boolean;
  /* if should skip mapping children, usually true when validating, false when using return */
  private skipMapper: boolean;
  /* internal counter of which child we are about to validate */
  private nextChildIndex = 0;

  constructor(
    componentClass: HasDisplayName,
    props: Object = { },
    children: ReactChild[],
    groups: IChildGroup[],
    throwError = false,
    skipMapper = false,
  ) {
    // Instead of React.ReactChild[], assume that child validation is only used with components
    // that have other components as children instead of functions (or other types).
    // TODO support validation of function as children and more
    this.children = children as ReactElement[];

    this.componentClass = componentClass
    this.props = props;
    this.groups = groups;
    this.throwError = throwError;
    this.skipMapper = skipMapper;
  }

  /*
   * Loops through all the expected child groups, validates, and returns the mapped result
   */
  matchChildren(): ChildGroupMatchReturn[] {
    const mapped = this.groups.map(childGroup => {
      const matched = this.match({
        ...childGroup,
        mapper: childGroup.mapper || (x => x),
      });

      if (childGroup.match.max <= 1) {
        // if looking for at most 1
        // return single element instead of array
        return matched[0];
      }

      // otherwise return array
      return matched;
    });

    /*
     * Once we've finished matching all groups
     * check if there are any children leftover
     */
    checkUnmatchedChildren(
      this.componentClass,
      this.nextChildIndex,
      this.children.length,
      this.throwError,
    );

    return mapped;
  }

  /*
   * Given a `childGroup`, keep matching children until one doesn't match or the
   * `max` is reached based on the provided range. When a child is matched apply
   * the mapper on the child
   */
  private match(childGroup: IChildGroup): ReactElement[] {
    const childrenMatched: ReactElement[] = [];

    while (
      this.nextChildIndex < this.children.length
      && (childGroup.match.max === undefined || childrenMatched.length < childGroup.match.max)
    ) {
      const child = this.children[this.nextChildIndex];

      if (validateChild(childGroup.match.types, child)) {
        const mappedChild = this.skipMapper
          ? child
          : childGroup.mapper(child, this.props);
        childrenMatched.push(mappedChild);
        this.nextChildIndex++;
      } else {
        // break out of the loop if doesn't match
        break;
      }
    }

    checkIncorrectChildrenCountError(
      this.componentClass,
      childGroup,
      childrenMatched.length,
      this.children[this.nextChildIndex],
      this.throwError,
    );

    return childrenMatched;
  }
}
