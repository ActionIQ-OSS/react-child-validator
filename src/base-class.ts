import * as React from "react";
import { IChildGroup, ChildGroupMatchReturn, ReactClass, ReactElement } from "./child-validator-types";
import { ReactChildValidator } from "./child-validator";

interface IBaseClass {
  childGroups?: IChildGroup[];
  propTypes?: {
    children: (props: {children: {}}, propName: string, componentName: string) => Error,
  };
  displayName: string;
}

type ClassConstructor = IBaseClass & ReactClass;

export abstract class BaseClass<P = {}, S = {}> extends React.PureComponent<P, S> {
  static childValidatorProps = (componentClass: IBaseClass) => ({
    children: (props: {children: {}}, _: string, componentName: string): Error => {
      try {
        new ReactChildValidator(
          componentClass,
          { },
          React.Children.toArray(props.children),
          componentClass.childGroups,
          true,
          true,
        ).matchChildren();
        return null;
      } catch (e) {
        return e;
      }
    },
  });

  matchChild(): ChildGroupMatchReturn {
    return this.matchChildren()[0];
  }

  matchChildren(): ChildGroupMatchReturn[] {
    return new ReactChildValidator(
      (this.constructor as ClassConstructor),
      this.props,
      React.Children.toArray(this.props.children),
      (this.constructor as ClassConstructor).childGroups,
      false,
      false,
    ).matchChildren();
  }
}

