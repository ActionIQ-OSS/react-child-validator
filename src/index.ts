import {
  zeroOrOneOf,
  zeroOrMoreOf,
  oneOf,
  oneOrMoreOf,
  countOf,
  countOrMoreOf,
  countOrLessOf,
  countBetweenOf,
  withMapper,
} from "./child-validator-utils";
import { BaseClass } from "./base-class";
import { ReactChildValidator } from "./child-validator";
import { ReactChild } from "./child-validator-types";

export {
  BaseClass,
  ReactChildValidator,

  zeroOrOneOf,
  zeroOrMoreOf,
  oneOf,
  oneOrMoreOf,
  countOf,
  countOrMoreOf,
  countOrLessOf,
  countBetweenOf,

  withMapper,

  ReactChild,
};