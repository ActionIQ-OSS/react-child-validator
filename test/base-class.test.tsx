import * as React from "react";
const { mount, render } = require("enzyme");
import { oneOf, zeroOrMoreOf, BaseClass, children } from "../src/";

class Element1 extends BaseClass<any, any> {
  static displayName = "Element1";

  render() {
    return <div>Test1</div>;
  }
}

class Element2 extends BaseClass<any, any> {
  static displayName = "Element2";

  render() {
    return <div>Test2</div>;
  }
}

class Element3 extends BaseClass<any, any> {
  static displayName = "Element3";

  render() {
    return <div>Test3</div>;
  }
}

class Wrapper1 extends BaseClass<any, any, typeof Wrapper1> {
  static displayName = "Wrapper1";

  static propTypes = BaseClass.childValidatorProps(Wrapper1);

  static childGroups = children(
    oneOf(Element1),
    oneOf(Element2),
    zeroOrMoreOf(Element3, Element2),
  );

  render() {
    const [el1, el2, el3] = this.matchChildren();

    return (
      <div>
        {el1}
        {el2}
        <div>
          {el3.map((el, i: number) => (
            <span key={i}>
              {el}
            </span>
          ))}
        </div>
      </div>
    );
  }
}

describe("BaseClass", () => {
  it("checks that a component's children are rendered properly", () => {
    expect(
      mount(
        <Wrapper1>
          <Element1 />
          <Element2 />
          <Element3 />
        </Wrapper1>,
      )
    ).toMatchSnapshot();

    expect(
      mount(
        <Wrapper1>
          <Element1 />
          <Element2 />
          <Element3 />
          <Element3 />
          <Element3 />
        </Wrapper1>,
      )
    ).toMatchSnapshot();

    expect(
      mount(
        <Wrapper1>
          <Element1 />
          <Element2 />
          <Element3 />
          <Element2 />
        </Wrapper1>,
      )
    ).toMatchSnapshot();
  });

  it("check that proptypes throw when children do not match", () => {
    Object.assign(global.console, { error: jest.fn() });

    mount(
      <Wrapper1>
        <Element1 />
      </Wrapper1>,
    );

    expect(console.error).toHaveBeenLastCalledWith(
      "Warning: Failed prop type: Too few children of type Element2 for <Wrapper1 />. Expected 1, Received: 0."
      + "\n    in Wrapper1",
    );
  });
})