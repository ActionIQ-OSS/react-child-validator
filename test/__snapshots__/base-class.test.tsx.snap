// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseClass checks that a component's children are rendered properly 1`] = `
<Wrapper1>
  <div>
    <Element1
      key=".0"
    >
      <div>
        Test1
      </div>
    </Element1>
    <Element2
      key=".1"
    >
      <div>
        Test2
      </div>
    </Element2>
    <div>
      <span
        key="0"
      >
        <Element3
          key=".2"
        >
          <div>
            Test3
          </div>
        </Element3>
      </span>
    </div>
  </div>
</Wrapper1>
`;

exports[`BaseClass checks that a component's children are rendered properly 2`] = `
<Wrapper1>
  <div>
    <Element1
      key=".0"
    >
      <div>
        Test1
      </div>
    </Element1>
    <Element2
      key=".1"
    >
      <div>
        Test2
      </div>
    </Element2>
    <div>
      <span
        key="0"
      >
        <Element3
          key=".2"
        >
          <div>
            Test3
          </div>
        </Element3>
      </span>
      <span
        key="1"
      >
        <Element3
          key=".3"
        >
          <div>
            Test3
          </div>
        </Element3>
      </span>
      <span
        key="2"
      >
        <Element3
          key=".4"
        >
          <div>
            Test3
          </div>
        </Element3>
      </span>
    </div>
  </div>
</Wrapper1>
`;

exports[`BaseClass checks that a component's children are rendered properly 3`] = `
<Wrapper1>
  <div>
    <Element1
      key=".0"
    >
      <div>
        Test1
      </div>
    </Element1>
    <Element2
      key=".1"
    >
      <div>
        Test2
      </div>
    </Element2>
    <div>
      <span
        key="0"
      >
        <Element3
          key=".2"
        >
          <div>
            Test3
          </div>
        </Element3>
      </span>
      <span
        key="1"
      >
        <Element2
          key=".3"
        >
          <div>
            Test2
          </div>
        </Element2>
      </span>
    </div>
  </div>
</Wrapper1>
`;
